Don't evaluate liquid for editing body / markdown part

jtkech opened the pull request:
---

So, this PR prevents from evaluating liquid while editing an html / markdown part

---

Then doing it i saw that body and markdown part don't evaluate liquid syntax for their `BodyAspect`. Maybe OK if the liquid contains query which may return many data. If so **maybe good to also remove the liquid evaluation in `LiquidPartHandler` for the `BodyAspect`**.

Otherwise, if better to evaluate liquid for `BodyAspect` **we could do it in the body and markdown handlers as it is done in the `LiquidPartHandler`**. But by taking care to create and register an `XxxPartViewModel` and set the `ContentItem` to the template context as we are used to do when rendering (not done in `LiquidPartHandler`), otherwise the evaluation may fail.

---

Finally, maybe the `LiquidPart` is useless if it is as an `HtmlPart` in standard editing that also allows liquid syntax. The only difference is that the liquid is evaluated in the `LiquidShapes` in place of the driver.

deanmarcussen commented:
@jtkech looks good, I had wondered why this liquid evaluation was always being called on Edit, but thought it was intentional.

For infos, and another PR (I'll open an issue) I think we should also update the editor to evaluate the liquid syntax to see if it is valid, like we do for example, in the `AutoroutePartSettingsDisplayDriver` 
i.e. 
```
if (!string.IsNullOrEmpty(model.Pattern) && !_templateManager.Validate(model.Pattern, out var errors))
{
    context.Updater.ModelState.AddModelError(nameof(model.Pattern), T["Pattern doesn't contain a valid Liquid expression. Details: {0}", string.Join(" ", errors)]);
} else {
...
}
```

jtkech commented:
I updated the  `BodyAspect` of Html, Markdown and Liquid part handlers.

I noticed that `GetContentItemAspectAsync` is called 2 times through `AspectsContentIndexHandler` and `FullTextAspectSettingsHandler`. I think because of `FullTextAspectSettings.IncludeBodyAspect`.

So, i added a scoped cache in the handler to not evaluate things multiple times.

Todo: See if the `BodyAspect` of `FacebookPluginPartHandler` needs some updates.

@deanmarcussen good idea, i will add the liquid validation, at least for the parts i updated in this PR, maybe another PR if any missing parts and for the fields.

jtkech commented:
- @deanmarcussen i added the **liquid validation** for the html, markdown, liquid and facebookPlugin parts, not done for fields, maybe in another PR.

- **Note**: This let me see some little issues **when a validation fails**. E.g for markdown the editor disappeared but i fixed it by calling the right `Edit()` in the driver. That said, for all the parts i tried, **the order of the content item fields changes, the Title go down**, to be fixed maybe in another PR.

- I also provided a more complete contextualization (e.g setting a model) before the liquid template manager is used in the related drivers / handlers. Here also only for the related parts, not done for fields, maybe in another PR.

deanmarcussen commented:
@jtkech sorry mate, I just make more work for you ;)

So I fix the problem with the field ordering on https://github.com/OrchardCMS/OrchardCore/pull/4724 

But looks great to do the liquid evaluation. Thanks!

hishamco commented on src/OrchardCore.Modules/OrchardCore.Markdown/Drivers/MarkdownBodyPartDisplay.cs:
```diff
@@ -14,34 +15,49 @@ public class MarkdownBodyPartDisplay : ContentPartDisplayDriver<MarkdownBodyPart
+            T = localizer;
         }
 
+        public IStringLocalizer T { get; private set; }
```
Same as above

jtkech replied:
done

jtkech commented:
Also fixes #2218

jtkech commented on src/OrchardCore.Modules/OrchardCore.Liquid/Drivers/LiquidPartDisplay.cs:
```diff
@@ -10,17 +11,16 @@ namespace OrchardCore.Liquid.Drivers
+            T = localizer;
         }
 
+        public IStringLocalizer T { get; private set; }
```
Okay

Note: it is done like this in 5 other places in the source code ;)

hishamco replied:
I will clean up everything, I don't think it's suitable anymore while the value set within the constructor

jtkech replied:
Yes, it's only set in the ctor and only for internal usage so we don't need a setter. I think we made it a property just to use an Uppercase `T` without breaking the naming convention.

I already removed in some places the `private` access modifier, not sure it's good but it is defined like this in most other places.

    public IStringLocalizer T { get; set; }

Oh yes, maybe you meant to remove the setter completely which makes more sense.

    public IStringLocalizer T { get; }

Or just this but that breaks the naming convention, but it doesn't hurt me, i already saw it in the aspnetcore code.

    private IStringLocalizer T;

**So i let you cleanup this as you want in another PR ;)**

But but but, maybe this is not the end of the story. As i remember for all `IStringLocalizer` it would be better to use `S` in place of `T`. And for all `IHtmlLocalizer` it would be better to use `H`.

jtkech commented on src/OrchardCore.Modules/OrchardCore.Liquid/Drivers/LiquidPartDisplay.cs:
```diff
@@ -38,7 +38,19 @@ public override IDisplayResult Edit(LiquidPart liquidPart)
+            {
+                if (!string.IsNullOrEmpty(viewModel.Liquid) && !_liquidTemplatemanager.Validate(viewModel.Liquid, out var errors))
+                {
+                    updater.ModelState.AddModelError(nameof(model.Liquid), T["The 'Body' contains an invalid Liquid expression. Details: {0}", string.Join(" ", errors)]);
```
We can't, it will be html encoded.

As i remember this is for the same reason that we moved some localizations from tag attributes values to tag contents. But here, we can't move the error message to another place.

Note: Seems that the liquid validation ends on the 1st error, as i have tried there is always only one error.

hishamco replied:
> We can't, it will be html encoded.

Hmm we did something similar in delete feature confirm dialog

jtkech commented:
@deanmarcussen 

> So I fix the problem with the field ordering on #4724

Good catch! Already integrated here

> But looks great to do the liquid evaluation. Thanks!

Okay, i think i will do it for fields, and also i will update the contextualization everywhere we call `liquidTemplateManager.RenderAsync()`.

sebastienros commented on src/OrchardCore.Modules/OrchardCore.Contents/Handlers/FullTextAspectSettingsHandler.cs:
```diff
@@ -72,12 +72,13 @@ public override Task GetContentItemAspectAsync(ContentItemAspectContext context)
                 if (settings.IncludeFullTextTemplate && !String.IsNullOrEmpty(settings.FullTextTemplate))
                 {
                     var templateContext = new TemplateContext();
+                    templateContext.SetValue("ContentItem", context.ContentItem);
```
Shouldn't we decide?

sebastienros replied:
I mean next line defines `Model`

jtkech replied:
You're right, it is just to allow to use `ContentItem` directly as in many other places.

In these other places it was the opposite we could only use `ContentItem` but not `Model.ContentItem`. By adding a model in these other places we can also use `Model.ContentItem`. So

- Do we need to remove the `SetValue("ContentItem")` everywhere because we now have the model? But it would break liquid expressions not using `Model.ContentItem`.

- Or let it as it is so that everywhere we can both use `ContentItem` and `Model.ContentItem`.

Note: There are some remaining places (as in workflows) where there is nothing to build a model so just a `ContentItem` is set. But i can build a model that only hold a `ContentItem` property, anyway i will do it and then also remove `SetValue("ContentItem")` depending on what we decide above.

jtkech replied:
Oh yes i remember, just below we define the `ContentItem` as the model, but because `ContentItem` has a `ContentItem` property pointing to itself, we can use `Model.ContentItem` as in other places, another singularity ;)

That's why i also added `ContentItem` to allow e.g `ContentItem.DisplayText` as we can do in other places.

I will remove it here.

sebastienros commented on src/OrchardCore.Modules/OrchardCore.Facebook/Widgets/Handlers/FacebookPluginPartHandler.cs:
```diff
@@ -21,10 +25,28 @@ public override Task GetContentItemAspectAsync(ContentItemAspectContext context,
+                    var templateContext = new TemplateContext();
+                    templateContext.SetValue("ContentItem", part.ContentItem);
+                    templateContext.MemberAccessStrategy.Register<FacebookPluginPartViewModel>();
+                    templateContext.LocalScope.SetValue("Model", model);
```
Don't use `LocalScope`

jtkech replied:
For infos this is what we do in the `LiquidViewTemplate.ContextualizeAsync()` when a model is provided by the `displayContext.Value`. And already done in some other places.

Do i still need to change it?

jtkech replied:
Just saw that `templateContext.SetValue("Model", model);` does exactly the same, i will replace it everywhere

jtkech replied:
Done everywhere

sebastienros commented on src/OrchardCore.Modules/OrchardCore.Html/Handlers/HtmlBodyPartHandler.cs:
```diff
@@ -1,34 +1,58 @@
+                {
+                    var model = new HtmlBodyPartViewModel()
+                    {
+                        Html = part.Html,
```
Fishy to apply the same property to both `Source` and `Html` ...

jtkech replied:
Yes, `Source` was an already existing property that i don't need anymore, i kept it in case someone already use it in a liquid expression. At this stage, before the liquid evaluation, `Source == Html`.

E.g the markdown view model (also the field as i remember) has a `Markdown` property and an `Html` property. The `LiquidPart` view model has a `Liquid` and an `Html` property. Here the singularity is that the `Html` property already exists. Note: the html field doesn't have a `Source` property.

**So if you want i can just remove this `Source` property**.

Note: Here normally we also don't need to pass the `Html` property that we want to evaluate, but i said to me that maybe it would be useful, so just pass the more complete model as we can.

jtkech replied:
Done, i removed the `Source` property

jtkech commented:
@sebastienros i reacted to your review

Just one thing, you asked me to remove `templateContext.SetValue("ContentItem", ...)` in `FullTextAspectSettingsHandler`, because we also define the `ContentItem` as the model.

Yes, because `ContentItem` has a `ContentItem` property pointing to itself, we can also use `Model.ContentItem` as in other places. That's why i also added `ContentItem` to also allow e.g `ContentItem.DisplayText` as we can do in other places.

Finally, having the `ContentItem` as the model also allows e.g `Model.DisplayText`, maybe too much, i agree, but i thought it would be good to also allow the expressions that we can use in other places.

So, i removed it as you asked me, but let me know if i need to reintroduce it OR maybe remove it in any other places? This to be consistent but it would be a breaking change.

jtkech commented:
Also Fixes #4736.

hishamco commented on src/OrchardCore.Modules/OrchardCore.Autoroute/Settings/AutoroutePartSettingsDisplayDriver.cs:
```diff
@@ -21,7 +21,7 @@ public AutoroutePartSettingsDisplayDriver(ILiquidTemplateManager templateManager
         }
 
-        public IStringLocalizer T { get; private set; }
+        public IStringLocalizer T { get; set; }
```
Why it's setter?

jtkech replied:
Maybe you didn't see my comment that you can see if you open below your `1 change requested` and go to `See review`. Here i copy pasted this comment. I agree we don't need this setter and **it can be removed in many other places**.

But you commented that you will do a cleanup PR for this, yes i think it is better because it is related to other files that are not changed here. Then  see my below comment where i also talk about `IStringLocalizer` and `IHtmlLocalizer` that could be renamed too and so on. 

So, here i just give you more infos and let you do it as you want in a separate PR.

---

Yes, it's only set in the ctor and only for internal usage so we don't need a setter. I think we made it a property just to use an Uppercase `T` without breaking the naming convention.

I already removed in some places the `private` access modifier, not sure it's good but it is defined like this in most other places.

    public IStringLocalizer T { get; set; }

Oh yes, maybe you meant to remove the setter completely which makes more sense.

    public IStringLocalizer T { get; }

Or just this but that breaks the naming convention, but it doesn't hurt me, i already saw it in the aspnetcore code.

    private IStringLocalizer T;

**So i let you cleanup this as you want in another PR ;)**

But but but, maybe this is not the end of the story. As i remember for all `IStringLocalizer` it would be better to use `S` in place of `T`. And for all `IHtmlLocalizer` it would be better to use `H`.

jtkech replied:
Okay, finally, as i am on thsi i will remove all setters, but without renaming the localizers ;)

hishamco replied:
Just leave it to me you can rebase on dev later

+1 for `private IStringLocalizer T;` coz removing a property for any reason later will be a breaking change

hishamco replied:
It will be better if we add `readonly` to the field :)

jtkech replied:
Too late, i changed it everywhere (88 files) it blowed up my PR, i hate you ;)

hishamco replied:
OMG, I just pushed a PR #4750, I told you leave it to me 😄 I know it blows

jtkech replied:
no problem, i will let you a comment on your new PR

jtkech commented on src/OrchardCore.Modules/OrchardCore.Title/Settings/TitlePartSettingsDisplayDriver.cs:
```diff
@@ -59,10 +59,10 @@ public override async Task<IDisplayResult> UpdateAsync(ContentTypePartDefinition
             else
             {
-                context.Builder.WithSettings(new TitlePartSettings {Pattern = model.Pattern, Options = model.Options,});
+                context.Builder.WithSettings(new TitlePartSettings { Pattern = model.Pattern, Options = model.Options, });
```
Okay, it was already there, i will remove it.

sebastienros commented:
Can you fix the conflicts and merge this PR (Squash merge, don't click "Retry" if it fails, reload the page)

sebastienros commented:
Fixes https://github.com/OrchardCMS/OrchardCore/issues/4736
